tests/tests_e3/job/job_controller_test.py

```diff
@@ -0,0 +1,199 @@
+from __future__ import absolute_import, division, print_function
+
+import sys
+
+from e3.anod.status import ReturnValue
+from e3.collection.dag import DAG
+from e3.job import EmptyJob, ProcessJob, StatusBasedJobController
+
+
+class ControlledJob(ProcessJob):
+    """A ProcessJob for testing purposes.
+
+    The process's behavior is configured by the suffix's UID (see
+    the "cmd" line method for more details on this.
+    """
+
+    def __init__(self, uid, data, notify_end):
+        super(ControlledJob, self).__init__(uid, data, notify_end)
+        self.run_count = 0
+
+    def run(self):
+        self.run_count += 1
+        return super(ControlledJob, self).run()
+
+    @property
+    def cmdline(self):
+        result = [sys.executable, '-c']
+        if self.uid.endswith('bad'):
+            result.append('import sys; sys.exit(1)')
+        elif self.uid.endswith('notready:once') and self.run_count < 2:
+            result.append('import sys; sys.exit(%d)'
+                          % ReturnValue.notready.value)
+        elif self.uid.endswith('notready:always'):
+            result.append('import sys; sys.exit(%d)'
+                          % ReturnValue.notready.value)
+        else:
+            result.append('print("Hello World")')
+        return result
+
+
+class SimpleController(StatusBasedJobController):
+    """A simple controller."""
+
+    def __init__(self, actions):
+        super(SimpleController, self).__init__(actions)
+        # The list of jobs (by UID) that have been requeued.
+        self.requeued = {}
+
+    def request_requeue(self, job):
+        """Requeue the job is not already queued once."""
+        # First record the number of times we've been asked to requeue
+        # that job, and allow requeuing only twice.
+        if job.uid not in self.requeued:
+            self.requeued[job.uid] = 0
+        self.requeued[job.uid] += 1
+        return self.requeued[job.uid] < 3
+
+    def create_job(self, uid, data, predecessors, notify_end):
+        if uid.endswith('dry-run'):
+            # Create a dry-run job, which is a job that never runs
+            # (EmptyJob) and returns ReturnValue.success.
+            return EmptyJob(uid, data, notify_end, ReturnValue.success)
+        else:
+            return ControlledJob(uid, data, notify_end)
+
+
+class TestStatusBasedJobController(object):
+
+    def test_good_job_no_predecessors(self):
+        """simple case of a leaf job."""
```

Comment:
Simple

Reply:
Fixed.

Reply:
Is it?

Reply:
Sorry. The interface makes it hard for me to verify that I fixed everything. I went through the list like 3 times to make sure, but ended up missing it anyway. New patch has been uploaded, and is currently being validated by the CB...